test/Microsoft.AspNet.Mvc.TagHelpers.Test/ScriptTagHelperTest.cs

```diff
@@ -0,0 +1,255 @@
+            var output = MakeTagHelperOutput("script");
+            var logger = new Logger();
+
+            // Act
+            var helper = new ScriptTagHelper
+            {
+                Logger = logger,
+            };
+
+            await helper.ProcessAsync(context, output);
+
+            // Assert
+            Assert.NotNull(output.TagName);
+            Assert.False(output.ContentSet);
+
+            Assert.Equal(1, logger.Logged.Count);
+
+            Assert.Equal(LogLevel.Verbose, logger.Logged[0].Item1);
+            Assert.IsAssignableFrom<ILoggerStructure>(logger.Logged[0].Item2);
+            Assert.StartsWith("Skipping processing for ScriptTagHelper",
+                ((ILoggerStructure)logger.Logged[0].Item2).Format());
+        }
+
+        [Fact]
+        public async Task PreservesOrderOfSourceAttributesWhenRun()
+        {
+            // Arrange
+            var context = MakeTagHelperContext(
+                attributes: new Dictionary<string, object>
+                {
+                    { "data-extra", "something"},
+                    { "src", "/blank.js"},
+                    { "data-more", "else"},
+                    { "asp-fallback-src", "http://www.example.com/blank.js" },
+                    { "asp-fallback-test", "isavailable()" },
+                });
+
+            var output = MakeTagHelperOutput("link",
+                attributes: new Dictionary<string, string>
+                {
+                    { "data-extra", "something"},
+                    { "src", "/blank.js"},
+                    { "data-more", "else"},
+                });
+
+            var logger = new Logger();
+
+            // Act
+            var helper = new ScriptTagHelper
+            {
+                Logger = logger,
+                FallbackSrc = "~/blank.js",
+                FallbackTestExpression = "http://www.example.com/blank.js",
+            };
+
+            await helper.ProcessAsync(context, output);
+
+            // Assert
+            Assert.StartsWith("<script data-extra=\"something\" src=\"/blank.js\" data-more=\"else\"", output.Content);
+            Assert.Equal(0, logger.Logged.Count);
+        }
+
+        private TagHelperContext MakeTagHelperContext(
+            IDictionary<string, object> attributes = null,
+            string content = null)
+        {
+            attributes = attributes ?? new Dictionary<string, object>();
+
+            return new TagHelperContext(attributes, Guid.NewGuid().ToString("N"), () => Task.FromResult(content));
+        }
+
+        private TagHelperOutput MakeTagHelperOutput(string tagName, IDictionary<string, string> attributes = null)
+        {
+            attributes = attributes ?? new Dictionary<string, string>();
+
+            return new TagHelperOutput(tagName, attributes);
+        }
+
+        private class Logger : ILogger<ScriptTagHelper>
+        {
+            public List<Tuple<LogLevel, object>> Logged { get; } = new List<Tuple<LogLevel, object>>();
```

Comment:
Make it a `List<LogMessage>` where `LogMessage` is a private class. It should be very small (properties and `Equal` override) and will enable you in all your log tests to create expected log messages instead of the hard to follow indexing that's happening right now.